### PR TITLE
[kubernetes] ignore empty endpoint changes

### PR DESCRIPTION
**Pull request by hensur**
### What does this PR do?
This PR filters out enpoint update events with no relevant changes in the `ResourceEventHandler`. I took most of the code from the coredns project, as they already implement this behaviour in the kubernetes plugin.

If an update for an enpoint comes in, and the Subset are equal in Addresses, Ports..., we can safely ignore the update. Traefik's config wouldn't change on a reload.

This affects endpoints like `kube-system/kube-controller-manager` that use an annotation (`control-plane.alpha.kubernetes.io/leader`) to perform leader election.

I also refactored the eventHandler code and moved it into a separate package, as it is shared between the `ingress` and `crd` clients.

### Motivation
We noticed updates in a one second interval to endpoints in `kube-system`, e.g. `kube-system/kube-controller-manager` or `kube-system/kube-scheduler`.

These updates caused traefik to reload the config every second, which created a noticeable performance impact. Traefik pods would "idle" using 600m of cpu shares. Setting the `ThrottleDuration` to a larger value would work around this problem, with the disadvantage that valid endpoint updates would be applied to the traefik config with a maximum delay of `ThrottleDuration`. This caused regular downtimes while pods were recreated.

### More

- [ ] Added/updated tests
- [ ] Added/updated documentation

### CODE REVIEW

**Comment by dtomcej** [comment]
Hello @hensur,

Thank you for contributing to the Traefik Project!

Since this is a performance update, is it possible to provide a before and after pprof graph, or some metrics graphs showing how this improvement works?

We would like to have some reference to work with when reviewing this code.

Thanks!

**Comment by SantoDE** [comment]
Hello @hensur,

do you have any references for us? We want to get further reviewing that PR :)

Thanks!

**Comment by hensur** [comment]
Hello,
to provide some further insights into the optimizations, I ran a cpu profile:

current traefik release (2.3.6), without any traffic:
```
Showing nodes accounting for 9.35s, 57.97% of 16.13s total
Dropped 440 nodes (cum <= 0.08s)
Showing top 20 nodes out of 171
      flat  flat%   sum%        cum   cum%
     1.81s 11.22% 11.22%      3.83s 23.74%  runtime.scanobject
     1.05s  6.51% 17.73%      3.85s 23.87%  runtime.mallocgc
     0.84s  5.21% 22.94%      1.02s  6.32%  runtime.findObject
     0.83s  5.15% 28.08%      0.83s  5.15%  hash/fnv.(*sum64).Write
     0.74s  4.59% 32.67%      0.74s  4.59%  runtime.memmove
     0.73s  4.53% 37.20%      0.83s  5.15%  runtime.heapBitsSetType
     0.50s  3.10% 40.30%      0.50s  3.10%  runtime.memclrNoHeapPointers
     0.42s  2.60% 42.90%      0.42s  2.60%  runtime.nextFreeFast
     0.41s  2.54% 45.44%      0.45s  2.79%  reflect.StructTag.Lookup
     0.30s  1.86% 47.30%      0.32s  1.98%  runtime.pageIndexOf (inline)
     0.24s  1.49% 48.79%      0.83s  5.15%  runtime.greyobject
     0.22s  1.36% 50.15%      0.22s  1.36%  runtime.markBits.isMarked (inline)
     0.21s  1.30% 51.46%      0.21s  1.30%  memeqbody
     0.17s  1.05% 52.51%      0.17s  1.05%  runtime.(*pallocBits).summarize
     0.17s  1.05% 53.56%      0.51s  3.16%  runtime.sweepone
     0.17s  1.05% 54.62%      0.39s  2.42%  strings.FieldsFunc
     0.15s  0.93% 55.55%         4s 24.80%  runtime.gcDrain
     0.15s  0.93% 56.48%      1.35s  8.37%  runtime.growslice
     0.12s  0.74% 57.22%      0.18s  1.12%  runtime.mapaccess1_faststr
     0.12s  0.74% 57.97%      0.28s  1.74%  runtime.mapassign_faststr
```

version compiled with these changes, that is also handling production traffic:
```
Showing nodes accounting for 440ms, 91.67% of 480ms total
Showing top 20 nodes out of 196
      flat  flat%   sum%        cum   cum%
      90ms 18.75% 18.75%      130ms 27.08%  github.com/traefik/traefik/v2/pkg/rules.host.func1
      50ms 10.42% 29.17%      190ms 39.58%  github.com/gorilla/mux.(*Route).Match
      50ms 10.42% 39.58%       50ms 10.42%  math/big.addMulVVW
      50ms 10.42% 50.00%       50ms 10.42%  syscall.Syscall
      20ms  4.17% 54.17%       30ms  6.25%  context.(*cancelCtx).Value
      20ms  4.17% 58.33%       40ms  8.33%  context.(*valueCtx).Value
      20ms  4.17% 62.50%       20ms  4.17%  runtime.adjustframe
      20ms  4.17% 66.67%       50ms 10.42%  runtime.mallocgc
      10ms  2.08% 68.75%       10ms  2.08%  crypto/sha256.block
      10ms  2.08% 70.83%       10ms  2.08%  encoding/json.stateEndValue
      10ms  2.08% 72.92%      140ms 29.17%  github.com/gorilla/mux.MatcherFunc.Match
      10ms  2.08% 75.00%       10ms  2.08%  net/http.http2validWireHeaderFieldName
      10ms  2.08% 77.08%       10ms  2.08%  net/url.unescape
      10ms  2.08% 79.17%       10ms  2.08%  runtime.(*mspan).refillAllocCache
      10ms  2.08% 81.25%       10ms  2.08%  runtime.(*randomOrder).start
      10ms  2.08% 83.33%       10ms  2.08%  runtime.(*spanSet).push
      10ms  2.08% 85.42%       10ms  2.08%  runtime.(*waitq).dequeueSudoG
      10ms  2.08% 87.50%       10ms  2.08%  runtime.efaceeq
      10ms  2.08% 89.58%       30ms  6.25%  runtime.gentraceback
      10ms  2.08% 91.67%       10ms  2.08%  runtime.getitab
```

As you can see, the current release will constantly allocate new objects, as it is scanning the current cluster state and reloading the config. This also leads to a higher cpu usage from garbage collection functions.

While the release version uses 16.13 cpu seconds in a 30s interval, the fixed version needs 480ms, while also handling production traffic.

[pprof_cpu.zip](https://github.com/traefik/traefik/files/5776756/pprof_cpu.zip)

**Review by dtomcej — APPROVED** [review]
LGTM
:shipit:

**Review by SantoDE — CHANGES_REQUESTED** [review]
Thanks for your work 👼 

Overall it looks awesome to me!

Im lacking a test tough. Do you think its feasible to add one?

**Comment by hensur** [comment]
@SantoDE I just pushed a test for this!

File: pkg/provider/kubernetes/ingress/client_test.go [hunk 1]
```diff
@@ -190,6 +191,102 @@ func TestClientIgnoresHelmOwnedSecrets(t *testing.T) {
...
+
+		assert.True(t, ep.Name == "empty-endpoint" || ep.Name == "filled-endpoint")
+	case <-time.After(50 * time.Millisecond):
+		assert.Fail(t, "expected to receive event for endpoints")
+	}
+
+	emptyEndpoint, err = kubeClient.CoreV1().Endpoints("test").Get(context.TODO(), "empty-endpoint", metav1.GetOptions{})
+	assert.NoError(t, err)
+
+	emptyEndpoint.Annotations["test-annotation"] = "___"
```
Comment (SantoDE) [inline pkg/provider/kubernetes/ingress/client_test.go:251]: Why do you change the annotation here? Why is this important?

Comment (hensur) [reply pkg/provider/kubernetes/ingress/client_test.go:251]: I wanted to simulate the changes to endpoints in my cluster that would previously trigger config reloads.
Those endpoints were used for leader election by updating the `control-plane.alpha.kubernetes.io/leader` annotation.
So this test will trigger an endpoint update without relevant changes. I want to make sure that these changes won't result in a new event in the event channel.

Comment (dtomcej) [reply pkg/provider/kubernetes/ingress/client_test.go:251]: Could we add a comment here like:

```
// Update endpoint annotation, but keep the resource version the same to show an update that should not trigger an update event on our eventCh.
```

Comment (hensur) [reply pkg/provider/kubernetes/ingress/client_test.go:254]: I added a slightly different comment, as the Resource Version is in fact updated (incremented from `1244` to `1245`).
I had to do this manually as the fake client doesn't seem to do it and the change detection logic i added will exit early if the resource version stays the same between updates.

**Review by SantoDE — APPROVED** [review]
LGTM. Thanks! 🎉

**Review by jbdoumenjou — APPROVED** [review]
LGTM :+1:

**Pull request merged**